[visionOS] Playing 360 or 180 videos displays an empty screen

jyavenard opened the pull request:
#### 2e2d6888bb5924a32e46209567cc1e52286c111a
<pre>
[visionOS] Playing 360 or 180 videos displays an empty screen
<a href="https://bugs.webkit.org/show_bug.cgi?id=287229">https://bugs.webkit.org/show_bug.cgi?id=287229</a>
<a href="https://rdar.apple.com/144206242">rdar://144206242</a>

Reviewed by Eric Carlson.

Detect if a vide&apos;s metadata set a projection kind and is one of:
- Equirectangular (projection of a 360 degree sphere to a rectangular plane)
- HalfEquirectangular (projection of a 180 degree hemisphere to a rectangular plane)
- ParametricImmersive (with FOV or fisheye)

And if so will pass this information to the LMK Player and will create an immersive entity instead of a planar one.

Manually tested on Vision Pro.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h: Adding new strings.
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::isImmersiveVideo const):
(WebCore::VideoTrackConfiguration::setIsImmersiveVideo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::isImmersiveVideoChanged):
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::maybeUpdateVideoMetadata):
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::spatialVideoMetadata const):
(WebCore::AVTrackPrivateAVFObjCImpl::isImmersiveVideo const):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WebCore::videoMetadataFromFormatDescription):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::updateConfiguration):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::PlaybackSessionInterfaceLMK::isImmersiveVideoChanged):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::isImmersiveVideoChanged):
(WebKit::PlaybackSessionManagerProxy::isImmersiveVideoChanged):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.maybeCreateSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.maybeClearSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.didCompleteExitFullscreen(_:any:)):
(WKSLinearMediaPlayer.makeDefaultEntity):
(WKSLinearMediaPlayer.maybeCreateSpatialEntity): Deleted.
(WKSLinearMediaPlayer.maybeClearSpatialEntity): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::isImmersiveVideoChanged):
(WebKit::PlaybackSessionManager::isImmersiveVideoChanged):

Canonical link: <a href="https://commits.webkit.org/290083@main">https://commits.webkit.org/290083@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92015f134818ab3a2e63f1cbb4dbcb1897641de5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8429 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93876 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39664 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16613 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68494 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26174 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91907 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6717 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48859 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6470 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38773 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76833 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/35745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95716 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16085 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11740 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77374 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16341 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76661 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18899 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21044 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9151 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16099 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15840 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19291 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17621 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
```diff
@@ -87,7 +87,13 @@
         }
 
-        if (auto* textTracks = oldMediaElement->audioTracks()) {
+        if (auto* videoTracks = oldMediaElement->videoTracks()) {
```
Isn’t this part of the other patch?

jyavenard replied:
it is, I've added the two patches to build in the PR. Need to limit to the top commit in the stack. I'll repush once the other land .

eric-carlson commented on Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
```diff
@@ -109,6 +115,11 @@
         audioTracks.addEventListener(events.changeEvent, *this, false);
         audioTracks.addEventListener(events.removetrackEvent, *this, false);
 
+        auto& videoTracks = newMediaElement->ensureVideoTracks();
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
```diff
@@ -214,30 +214,42 @@ std::optional<SpatialVideoMetadata> videoMetadataFromFormatDescription(CMFormatD
+        return spatialVideoMetadata;
+
+    auto checkForImmersiveData = [&]() -> bool {
+        auto projectionKind = dynamic_cf_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, CFSTR("ProjectionKind")));
```
I think you can use `PAL::kCMFormatDescriptionExtension_ProjectionKind` here

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
```diff
@@ -214,30 +214,42 @@ std::optional<SpatialVideoMetadata> videoMetadataFromFormatDescription(CMFormatD
+
+    auto checkForImmersiveData = [&]() -> bool {
+        auto projectionKind = dynamic_cf_cast<CFStringRef>(PAL::CMFormatDescriptionGetExtension(formatDescription, CFSTR("ProjectionKind")));
+        return projectionKind && (CFStringCompare(projectionKind, CFSTR("Equirectangular"), 0) == kCFCompareEqualTo || CFStringCompare(projectionKind, CFSTR("HalfEquirectangular"), 0) == kCFCompareEqualTo || CFStringCompare(projectionKind, CFSTR("ParametricImmersive"), 0) == kCFCompareEqualTo);
```
You should add `kCMFormatDescriptionProjectionKind_Equirectangular` and `kCMFormatDescriptionProjectionKind_HalfEquirectangular` to CoreMediaSoftLink

jyavenard replied:
could have sworn they weren't in the public SDK when I looked last !

jyavenard committed:
[visionOS] Playing 360 or 180 videos displays an empty screen
https://bugs.webkit.org/show_bug.cgi?id=287229
rdar://144206242

Reviewed by Eric Carlson.

Detect if a vide's metadata set a projection kind and is one of:
- Equirectangular (projection of a 360 degree sphere to a rectangular plane)
- HalfEquirectangular (projection of a 180 degree hemisphere to a rectangular plane)
- ParametricImmersive (with FOV or fisheye)

And if so will pass this information to the LMK Player and will create an immersive entity instead of a planar one.

Manually tested on Vision Pro.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h: Adding new strings.
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::toJSON const):
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::isImmersiveVideo const):
(WebCore::VideoTrackConfiguration::setIsImmersiveVideo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::isImmersiveVideoChanged):
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::maybeUpdateVideoMetadata):
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::spatialVideoMetadata const):
(WebCore::AVTrackPrivateAVFObjCImpl::isImmersiveVideo const):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WebCore::videoMetadataFromFormatDescription):
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::updateConfiguration):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::PlaybackSessionInterfaceLMK::isImmersiveVideoChanged):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::isImmersiveVideoChanged):
(WebKit::PlaybackSessionManagerProxy::isImmersiveVideoChanged):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.maybeCreateSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.maybeClearSpatialOrImmersiveEntity):
(WKSLinearMediaPlayer.didCompleteExitFullscreen(_:any:)):
(WKSLinearMediaPlayer.makeDefaultEntity):
(WKSLinearMediaPlayer.maybeCreateSpatialEntity): Deleted.
(WKSLinearMediaPlayer.maybeClearSpatialEntity): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::isImmersiveVideoChanged):
(WebKit::PlaybackSessionManager::isImmersiveVideoChanged):

Canonical link: https://commits.webkit.org/290083@main
